File: behavioral/iterator/data_stream_processor/typescript/data_stream_processor.ts
// Iterator Interface
export interface IIterator<T> {
    next(): { value: T | undefined; done: boolean };
    hasNext(): boolean;
}

// Aggregate Interface
export interface IAggregate<T> {
    createIterator(): IIterator<T>;
}

// Concrete Iterator
export class StreamIterator<T> implements IIterator<T> {
    private stream: DataStream<T>;
    private position: number = 0;

    constructor(stream: DataStream<T>) {
        this.stream = stream;
    }

    next(): { value: T | undefined; done: boolean } {
        if (!this.hasNext()) {
            return { value: undefined, done: true };
        }
        const value = this.stream.get(this.position);
        this.position++;
        return { value: value, done: false };
    }

    hasNext(): boolean {
        return this.position < this.stream.getCount();
    }
}

// Concrete Aggregate
export class DataStream<T> implements IAggregate<T> {
    private dataChunks: T[] = [];

    addChunk(chunk: T): void {
        this.dataChunks.push(chunk);
    }

    get(index: number): T {
        if (index < 0 || index >= this.dataChunks.length) {
            throw new Error("Index out of bounds");
        }
        return this.dataChunks[index];
    }

    getCount(): number {
        return this.dataChunks.length;
    }

    createIterator(): IIterator<T> {
        return new StreamIterator<T>(this);
    }
}

File: behavioral/iterator/data_stream_processor/python/data_stream_processor.py
from collections.abc import Iterator, Iterable


class StreamIterator(Iterator):
    """Concrete Iterator for the DataStream."""
    _position: int = None
    _stream: 'DataStream' = None

    def __init__(self, stream: 'DataStream'):
        self._stream = stream
        self._position = 0

    def __next__(self):
        """Returns the next element from the stream."""
        try:
            value = self._stream.get(self._position)
            self._position += 1
            return value
        except IndexError:
            raise StopIteration()


class DataStream(Iterable):
    """Concrete Aggregate that stores data chunks and provides an iterator."""

    def __init__(self):
        self._data_chunks = []

    def add_chunk(self, chunk: any) -> None:
        """Adds a data chunk to the stream."""
        self._data_chunks.append(chunk)

    def get(self, index: int) -> any:
        """Gets a data chunk at a specific index."""
        return self._data_chunks[index]

    def __len__(self) -> int:
        """Returns the total number of chunks in the stream."""
        return len(self._data_chunks)

    def __iter__(self) -> StreamIterator:
        """Returns the iterator for the stream."""
        return StreamIterator(self)